File: src/configuration.ts
import processorVanilla from './processOrderVanilla'
import processorVanillaSync from './processOrderVanillaSync'
import processorFp from './processOrderFp'
import processorFpChecked from './processOrderFpChecked'
import { categorizedOrderIds } from './data'
import { SyncProcessor, AsyncProcessor } from './api'
import { readFileSync } from 'fs'

const config = JSON.parse(readFileSync(__dirname + '/params.json', 'utf8')) as {
  warmup: number
  epoch: number
  failureRate: number
}

function getAsyncProcessor(processorName: string): AsyncProcessor | null {
  if (processorName === 'vanilla') return processorVanilla
  if (processorName === 'fp') return processorFp
  if (processorName === 'checked') return processorFpChecked
  return null
}
function getSyncProcessor(processorName: string): SyncProcessor | null {
  if (processorName === 'syncv') return processorVanillaSync
  //if (processorName === 'syncfp') return porcessorFp
  return null
}

export type BenchmarkIds = { ok: string[]; ko: string[] }

export type SyncBenchmarkConfiguration = {
  isSync: true
  name: string
  processor: SyncProcessor
  warmup: number
  failureRate: number
  epoch: number
}
export type AsyncBenchmarkConfiguration = {
  isSync: false
  name: string
  processor: AsyncProcessor
  warmup: number
  failureRate: number
  epoch: number
}
export type BenchmarkConfiguration =
  | SyncBenchmarkConfiguration
  | AsyncBenchmarkConfiguration

export function get(): BenchmarkConfiguration {
  const processorName = process.argv[2]

  let syncProcessor = getSyncProcessor(processorName)
  if (syncProcessor !== null) {
    return {
      isSync: true,
      name: processorName,
      processor: syncProcessor,
      warmup: config.warmup,
      epoch: config.epoch,
      failureRate: config.failureRate
    }
  }

  let asyncProcessor = getAsyncProcessor(processorName)
  if (asyncProcessor !== null) {
    return {
      isSync: false,
      name: processorName,
      processor: asyncProcessor,
      warmup: config.warmup,
      epoch: config.epoch,
      failureRate: config.failureRate
    }
  }

  throw new Error('Processor not recognized')
}

export type RunnerResult = {
  ok_counter: number
  ko_counter: number
  total: number
}

export function syncRunner(
  processor: SyncProcessor,
  iterations: number,
  failure_rate: number,
  ids: BenchmarkIds
): RunnerResult {
  let ok_counter = 0
  let ko_counter = 0
  let total = 0.0

  while (ok_counter + ko_counter < iterations) {
    let id = ''
    if (ok_counter > 0 && ko_counter / ok_counter < failure_rate) {
      id = ids.ko[ko_counter % ids.ko.length]
      ko_counter += 1
    } else {
      id = ids.ok[ok_counter % ids.ok.length]
      ok_counter += 1
    }
    total += processor(id).totalAmount
  }
  return {
    ok_counter,
    ko_counter,
    total
  }
}

export async function asyncRunner(
  processor: AsyncProcessor,
  iterations: number,
  failure_rate: number,
  ids: BenchmarkIds
): Promise<RunnerResult> {
  let ok_counter = 0
  let ko_counter = 0
  let total = 0.0

  while (ok_counter + ko_counter < iterations) {
    let id = ''
    if (ok_counter > 0 && ko_counter / ok_counter < failure_rate) {
      id = ids.ko[ko_counter % ids.ko.length]
      ko_counter += 1
    } else {
      id = ids.ok[ok_counter % ids.ok.length]
      ok_counter += 1
    }
    total += (await processor(id)).totalAmount
  }
  return {
    ok_counter,
    ko_counter,
    total
  }
}

export async function benchmark(
  config: BenchmarkConfiguration
): Promise<[number, RunnerResult]> {
  if (config.isSync) {
    syncRunner(
      config.processor,
      config.warmup,
      config.failureRate,
      categorizedOrderIds
    )
    const start = new Date().getTime()
    let result = syncRunner(
      config.processor,
      config.epoch,
      config.failureRate,
      categorizedOrderIds
    )
    const end = new Date().getTime()
    return [end - start, result]
  } else {
    await asyncRunner(
      config.processor,
      config.warmup,
      config.failureRate,
      categorizedOrderIds
    )
    const start = new Date().getTime()
    let result = await asyncRunner(
      config.processor,
      config.epoch,
      config.failureRate,
      categorizedOrderIds
    )
    const end = new Date().getTime()
    return [end - start, result]
  }
}

File: rust/src/configuration.rs
use crate::api::*;
use crate::data::{get_order, ORDERS};
use crate::process_order_vanilla::VanillaProcessOrder;
use async_trait::async_trait;
use std::time::{Duration, Instant};

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum ProcessorKind {
    Vanilla,
    Fp,
    Null,
}

struct NullProcessOrder {}

#[async_trait]
impl ProcessOrder for NullProcessOrder {
    async fn process(&self, order_id: &String) -> PlaceOrderResult {
        let order = get_order(order_id);
        match order {
            None => Err(OrderNotValid::NoItems),
            Some(o) => match validate_order(o) {
                Ok(_) => Ok(OrderSuccessful::new(0.0)),
                Err(err) => Err(err),
            },
        }
    }
}

#[async_trait]
impl Processor for NullProcessOrder {
    async fn process(&self, _order_id: &String) -> () {
        ()
    }
}

impl NullProcessOrder {
    pub fn process_order() -> &'static dyn ProcessOrder {
        &(NullProcessOrder {}) as &dyn ProcessOrder
    }
    pub fn processor() -> &'static dyn Processor {
        &(NullProcessOrder {}) as &dyn Processor
    }
}

impl ProcessorKind {
    pub fn get_process_order(&self) -> &dyn ProcessOrder {
        match self {
            ProcessorKind::Vanilla => VanillaProcessOrder::process_order(),
            ProcessorKind::Fp => NullProcessOrder::process_order(),
            ProcessorKind::Null => NullProcessOrder::process_order(),
        }
    }

    pub fn get_processor(&self) -> &dyn Processor {
        match self {
            ProcessorKind::Vanilla => VanillaProcessOrder::processor(),
            ProcessorKind::Fp => NullProcessOrder::processor(),
            ProcessorKind::Null => NullProcessOrder::processor(),
        }
    }

    pub fn name(&self) -> &'static str {
        match self {
            ProcessorKind::Vanilla => "vanilla",
            ProcessorKind::Fp => "fp",
            ProcessorKind::Null => "null",
        }
    }
}

pub struct BenchmarkIds {
    pub ok: Vec<String>,
    pub ko: Vec<String>,
}

pub struct BenchmarkConfiguration {
    pub kind: ProcessorKind,
    pub warmup: i32,
    pub failure_rate: f64,
    pub epoch: i32,
    pub ids: BenchmarkIds,
}

impl BenchmarkConfiguration {
    pub fn process_order(&self) -> &dyn ProcessOrder {
        self.kind.get_process_order()
    }

    pub fn processor(&self) -> &dyn Processor {
        self.kind.get_processor()
    }

    pub fn name(&self) -> &'static str {
        self.kind.name()
    }

    pub fn report(&self, duration: Duration) {
        println!(
            "{}\t duration {}\t\t warmup {}\t iter {}\t failure {}",
            self.name(),
            duration.as_millis(),
            self.warmup,
            self.epoch,
            self.failure_rate
        );
    }

    pub async fn run(&self) {
        let duration = benchmark(self).await;
        self.report(duration);
    }
}

pub async fn get_configuration(kind: ProcessorKind) -> BenchmarkConfiguration {
    let mut ids = BenchmarkIds {
        ok: vec![],
        ko: vec![],
    };

    let process_order = kind.get_process_order();
    for id in ORDERS.keys() {
        match process_order.process(id).await {
            Ok(_) => ids.ok.push(id.clone()),
            Err(_) => ids.ko.push(id.clone()),
        }
    }

    BenchmarkConfiguration {
        kind,
        warmup: 1000000,
        failure_rate: 0.01,
        epoch: 1000000,
        ids,
    }
}

pub async fn runner(
    processor: &dyn Processor,
    iterations: usize,
    failure_rate: f64,
    ids: &BenchmarkIds,
) {
    let mut ok_counter: usize = 0;
    let mut ko_counter: usize = 0;

    while ok_counter + ko_counter < iterations {
        let id = if ok_counter > 0 && (ko_counter as f64) / (ok_counter as f64) < failure_rate {
            let id = &ids.ko[ko_counter % ids.ko.len()];
            ko_counter += 1;
            id
        } else {
            let id = &ids.ok[ok_counter % ids.ok.len()];
            ok_counter += 1;
            id
        };
        processor.process(id).await;
    }
}

pub async fn benchmark(config: &BenchmarkConfiguration) -> Duration {
    runner(
        config.processor(),
        config.warmup as usize,
        config.failure_rate,
        &config.ids,
    )
    .await;
    let start = Instant::now();
    runner(
        config.processor(),
        config.epoch as usize,
        config.failure_rate,
        &config.ids,
    )
    .await;
    start.elapsed()
}
